### DISCUSSION TITLE
Best way to track elapsed time between IRQs

### DISCUSSION THREAD

**Discussion by rm5248** [zephyrproject-rtos/zephyr#78458] [Q&A]
I'm trying to use Zephyr to decode a DCC signal.  I have my GPIO setup with an IRQ on both rising and falling edges.  In order to decode the signal, I need to know how many microseconds have elapsed between the IRQs.

A half-bit is 58 +- 6 microseconds long, so any work that needs to be done needs to be accurate to that level.

My simplified code looks like the following:

```
uint32_t prev_cycle;

static void gpio_pin_change(const struct device *dev, struct gpio_callback *cb, uint32_t pins){
	const uint32_t curr_cycle = k_cycle_get_32();
	const uint32_t cycle_time_usec = k_cyc_to_us_floor32(curr_cycle - prev_cycle);

	prev_cycle = curr_cycle;

        // Do something with cycle_time_usec to decode the bit
}
```

However, this gives inconsistent results where sometimes the cycle time is less than 52 microseconds, and sometimes more than 64 microseconds, when the data coming in is fine(as far as I can tell with my oscilloscope).

Thoughts on what I could do:

1. Write my own IRQ handler to remove any latency that the GPIO handler might be adding in
2. Increase the frequency of the processor.  I'm not sure what I'm running at now, but one of my theories is that the processor is not running fast enough to have the cycle time update correctly.
3. Use a secondary monotonic timer of some kind that counts microseconds.

Is one of the above ideas a reasonable thing to try?  Or is there something that I'm missing?

**Comment by eiten** [comment]
Well, best would be to use a hardware timer/counter. Which platform are you on?

**Reply by rm5248** [reply]
STM32.  I haven't found an example of using a hardware timer with zephyr, so I'm unsure where to start

**Comment by eiten** [comment]
Hm, did you check samples/drivers/misc/timeaware_gpio?
